libtransmission/peer-msgs.cc

```diff
@@ -1187,11 +1191,16 @@ void tr_peerMsgsImpl::parse_ltep_handshake(MessageReader& payload)
     auto var = tr_variant_serde::benc().inplace().parse(handshake_sv);
     if (!var || !var->holds_alternative<tr_variant::Map>())
     {
-        logtrace(this, "GET  extended-handshake, couldn't get dictionary");
+        logtrace(this, "got ltep handshake, couldn't get dictionary");
         return;
     }
 
-    logtrace(this, fmt::format("here is the base64-encoded handshake: [{:s}]", tr_base64_encode(handshake_sv)));
+    logtrace(this, fmt::format("got ltep handshake, base64-encoded body: [{:s}]", tr_base64_encode(handshake_sv)));
+
+    if (!io_->supports_ltep())
```

Comment:
I haven't found a client yet that does this across >100 clients. still think we should reject these clients when we see them rather than tolerating bad peers. spider-sense says tolerating will come back to bite us, though I can't spell out that scenario yet.

Reply:
No, look, it finally clicked with me, why the code is the way it is. Took me long enough...

The reserved byte flag is literally just a *hint* for the receiving client to send an ltep handshake on establishing the connection. In other words, all it does is to assure the receiving client that its ltep handshake won't fall on deaf ears. This part corresponds to the `send_ltep_handshake()` call in the constructor.

Even if the peer forgot to set the flag in the reserved byte, as long it sends out an ltep handshake, then the client can reasonably assume that the peer supports ltep, and reply with its own ltep handshake in case it didn't yet. This is where the `send_ltep_handshake()` call that we thought were redundant comes in. Or in other words, it is not actually redundant. In fact, it needs to be moved out of the `if (io->supports_ltep())` block.

* * *

I think we don't need to worry too much about pitfalls surrounding this flag.

It's not like we are sending around sensitive information with ltep, so there's no risk of accidentally leaking stuff. It's not like there's any control flow to ltep after the initial handshake, so there's no risk of falling into an "invalid state" either. Where can the risk be? The most harm it can do is waste some bandwidth.

Reply:
> No, look, it finally clicked with me, why the code is the way it is. Took me long enough...
> 
> The reserved byte flag is literally just a _hint_ for the receiving client to send an ltep handshake on establishing the connection. In other words, all it does is to assure the receiving client that its ltep handshake won't fall on deaf ears. This part corresponds to the `send_ltep_handshake()` call in the constructor.... in fact, it needs to be moved out of the if (io->supports_ltep()) block.

Hmm, interesting. What makes you think that? As I said, I haven't found a client that does that, but maybe I'm not thinking properly about the cases of clients that don't set LTEP flag but _would_ actually send LTEP (which, again, I haven't seen) in response to us sending.

Reply:
My point is, LTEP does not depend on the LTEP flag. Given the same LTEP message, the value of the flag does not change the message's meaning one bit.

If you think about it, it's the extended message IDs that are the crucial part of the protocol. All LTEP needs to function is for one side of the connection to send out its LTEP handshake first, and for the other side to parse the handshake and respond accordingly.

Having said that, it's still a bug if any client does not set the LTEP flag but sends LTEP, so it's nice to see that no client actually does that.